kv_storage.h

```diff
@@ -0,0 +1,383 @@
+    std::promise<void> pr;
+    std::future<void> future = pr.get_future();
+    if (listener_.storage_.data.count(entry.key()) && !T_POLICY::allow_overwrite_on_add) {
+      pr.set_exception(std::make_exception_ptr(SpecificEntryKeyAlreadyExistsException<T_ENTRY>(entry)));
+    } else {
+      // The order of the next two calls is important for eventual consistency.
+      listener_.mq_.EmplaceMessage(new MQMessageAdd(entry, std::move(pr)));
+      stream_.Publish(entry);
+    }
+    return future;
+  }
+
+  void AsyncAdd(const T_ENTRY& entry, T_VOID_CALLBACK on_done, T_VOID_CALLBACK on_exists) {
+    if (!listener_.storage_.data.count(entry.key())) {
+      // The order of the next two calls is important for eventual consistency.
+      listener_.mq_.EmplaceMessage(new MQMessageAdd(entry, on_done));
+      stream_.Publish(entry);
+    } else {
+      if (T_POLICY::allow_overwrite_on_add) {
+       listener_.mq_.EmplaceMessage(new MQMessageAdd(entry, on_exists));
+       stream_.Publish(entry);
+      } else {
+        std::async(on_exists);
+      }
+    }
+  }
+
+  void Add(const T_ENTRY& entry) { AsyncAdd(entry).get(); }
+
+  // For testing purposes.
+  bool CaughtUp() const { return listener_.caught_up_; }
+  size_t EntriesSeen() const { return listener_.entries_seen_; }
+
+ private:
+  // Stateful storage.
+  struct Storage {
+    std::unordered_map<T_KEY, T_ENTRY, typename T_KEY::HashFunction> data;
+  };
+
+  // The logic to "interleave" updates from Sherlock stream with inbound KeyValueStorage requests.
+  struct MQMessage {
+    virtual void DoIt(Storage& storage) = 0;
+  };
+
+  struct MQMessageEntry : MQMessage {
+    // TODO(dkorolev): A single entry is fine to copy, polymorphic ones should be std::move()-d.
+    T_ENTRY entry;
+
+    explicit MQMessageEntry(const T_ENTRY& entry) : entry(entry) {}
+
+    virtual void DoIt(Storage& storage) { storage.data[entry.key()] = entry; }
+  };
+
+  struct MQMessageGet : MQMessage {
+    T_KEY key;
+    std::promise<T_ENTRY> pr;
+    T_ENTRY_CALLBACK on_found;
+    T_KEY_CALLBACK on_not_found;
+
+    explicit MQMessageGet(const T_KEY& key, std::promise<T_ENTRY>&& pr)
+        : key(key), pr(std::move(pr)) {}
+    explicit MQMessageGet(const T_KEY& key, T_ENTRY_CALLBACK on_found, T_KEY_CALLBACK on_not_found)
+        : key(key), on_found(on_found), on_not_found(on_not_found) {}
+
+    virtual void DoIt(Storage& storage) {
+      const auto cit = storage.data.find(key);
+      if (cit != storage.data.end()) {
+        if (on_found) {  // Callback function defined.
+          std::async(on_found, cit->second);
+        } else {
+          pr.set_value(cit->second);
+        }
+      } else {  // Key not found.
+       if (on_not_found) {  // Callback function defined.
+          std::async(on_not_found, key);
+        } else if (T_POLICY::allow_nonthrowing_get) {  // Return non-existing entry.
+          T_ENTRY null_entry(CreateNullEntry<T_ENTRY, std::is_base_of<Nullable, T_ENTRY>::value>());
```

Comment:
Nit: Dot notation would he great here, if `set_key()` returns `*this`:

`pr.set_value(CreateNullEntry<T_ENTRY>().set_key(key));`.

Also, why isn't the `std::is_base_of<Nullable, T_ENTRY>::value` magic part of the `CreateNullEntry<>` logic?